### DISCUSSION TITLE
How to handle related documents?

### DISCUSSION THREAD

**Discussion by drewlyton** [automerge/automerge#695] [General]
Hey friends 👋!

Just stumbled across the [Local-first](https://www.inkandswitch.com/local-first/#developer-infrastructure-for-building-apps) article and Automerge thanks to a podcast I was listening to talking about [Linear](linear.app) and [Replicache](https://replicache.dev/). SO EXCITED by this project.

In reading through the documentation, watching talks and demos, it was unclear to me how to handle creating relational data with Automerge documents. I found [this page](https://automerge.org/docs/cookbook/modeling-data/) in the docs, but it didn't seem to answer the question about how to model relationships or even when to make a new document.

As a tool for discussion, imagine I'm working on a shared wiki application like [Slab](slab.com) or [Confluence](https://www.atlassian.com/software/confluence). Each Post could be related to one or many Posts as a Link (outbound connection) or Backlink (inbound connection). How could that be modeled with Automerge?

Also, what would be the tradeoffs to implementing a graph-database structure within the context of Automerge where every node/attribute in the graph is its own document keeping track of its own changes/history?

Thanks!

**Comment by ept** [comment]
Hi @drewlyton, sorry for the slow reply. You can model the data in the same way as you'd probably model the data in JSON: give each post a unique ID, and store them in a map from ID to post content (like a primary key index). However, Automerge currently doesn't provide secondary indexes, so if you want bidirectional links between posts, there are a few options:

* Store the links in a denormalised way, on both the outgoing and on the incoming side. However, as always with denormalisation, there is a risk of the two sides of the relationship becoming inconsistent with each other.
* Store the links in a normalised way, and build your own secondary index in a separate data structure outside of the Automerge document by [listening to patches](https://automerge.org/docs/tsapi/index/#listening-to-patches). For example, you could have a patch callback that looks for outgoing links being added or removed, and which correspondingly adds or removes the reverse link in your index.

Does that answer your question?

**Comment by revington** [comment]
Hey the listening to patches link is broken. Documentation does not tell how exactly can be used. 
Digging in the code I have seen that you can use this as an option on the `init()` function and also on the `changeDoc` like

```typescript

//https://github.com/automerge/automerge/blob/bf473308863b91ad24863e1a582bd83f36f9b211/javascript/src/stable.ts#L349
doc1 = automerge.change(doc1, {message: "add another value", time: 1640995200, patchCallback}, d => {
   d.key2 = "value2"
})
 ```
